tests/langchain/test_langchain_tracer.py

```diff
@@ -541,3 +543,48 @@ def worker_function(worker_id):
     traces = get_traces()
     assert len(traces) == 10
     assert all(len(trace.data.spans) == 1 for trace in traces)
+
+
+def test_tracer_does_not_add_spans_to_trace_after_root_run_has_finished(clear_singleton):
+    from langchain.callbacks.manager import CallbackManagerForLLMRun
+    from langchain.chat_models.base import SimpleChatModel
+    from langchain.schema.messages import BaseMessage
+
+    class FakeChatModel(SimpleChatModel):
+        """Fake Chat Model wrapper for testing purposes."""
+
+        def _call(
+            self,
+            messages: List[BaseMessage],
+            stop: Optional[List[str]] = None,
+            run_manager: Optional[CallbackManagerForLLMRun] = None,
+            **kwargs: Any,
+        ) -> str:
+            return TEST_CONTENT
+
+        @property
+        def _llm_type(self) -> str:
+            return "fake chat model"
+
+    run_id_for_on_chain_end = None
+
+    class ExceptionCatchingTracer(MlflowLangchainTracer):
+        def on_chain_end(self, outputs, *, run_id, inputs=None, **kwargs):
+            nonlocal run_id_for_on_chain_end
+            run_id_for_on_chain_end = run_id
+            super().on_chain_end(outputs, run_id=run_id, inputs=inputs, **kwargs)
+
+    prompt = SystemMessagePromptTemplate.from_template("You are a nice assistant.") + "{question}"
+    chain = prompt | FakeChatModel() | StrOutputParser()
+
+    tracer = ExceptionCatchingTracer()
+
+    chain.invoke(
+        "What is MLflow?",
+        config={"callbacks": [tracer]},
+    )
+
+    with pytest.raises(MlflowException, match="Span for run_id .* not found."):
+        # After the chain is invoked, verify that the tracer no longer holds references to spans,
+        # ensuring that the tracer does not add spans to the trace after the root run has finished
+        tracer.on_chain_end({"output": "test output"}, run_id=run_id_for_on_chain_end, inputs=None)
```

Comment:
On master, this fails with `exceptions.MlflowException: Span with ID 0x1e7032990cf7a762 is not found or already finished.`, indicating that the tracer is trying to add a span to a trace that is no longer active because the root run has already completed